rllib/models/specs/checker.py

```diff
@@ -0,0 +1,290 @@
+from collections import abc
+import functools
+from typing import Union, Mapping, Any, Callable
+
+from ray.util.annotations import DeveloperAPI
+
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.models.specs.specs_base import Spec, TypeSpec
+from ray.rllib.models.specs.specs_dict import SpecDict
+from ray.rllib.models.specs.typing import SpecType
+
+
+def _convert_to_canonical_format(spec: SpecType) -> Union[Spec, SpecDict]:
+    """Converts a spec type input to the canonical format.
+
+    The canonical format is either a tree structure with SpecType leaves (i.e.
+    SpecDict) or if the spec is a single constraint, a single SpecType object.
+
+    The input can be any of the following:
+        - a list of nested_keys. nested_keys are either strings or tuples of strings
+        specifying the path to a leaf in the tree.
+        - a tree of constraints. The tree structure can be specified as any nested
+        hash-map structure (e.g. dict, SpecDict, NestedDict, etc.) The leaves of the
+        tree can be either a Spec object, a type, or None. If the leaf is a type, it is
+        converted to a TypeSpec. If the leaf is None, only the existance of the key is
+        checked.
+        - a single constraint. The constraint can be a Spec object, a type, or None.
+
+    Args:
+        spec: The spec to convert to canonical format.
+
+    Returns:
+        The canonical format of the spec.
+    """
+    # convert spec of form list of nested_keys to model_spec with None leaves
+    if isinstance(spec, list):
+        spec = [(k,) if isinstance(k, str) else k for k in spec]
+        return SpecDict({k: None for k in spec})
+
+    # convert spec of form tree of constraints to model_spec
+    if isinstance(spec, abc.Mapping):
+        spec = SpecDict(spec)
+        for key in spec:
+            # if values are types or tuple of types, convert to TypeSpec
+            if isinstance(spec[key], (type, tuple)):
```

Comment:
So if `spec[key]` is None, we don't do anything?

Reply:
I'm still getting into this, but the docstring says that we check for existence. Where do we do that?

Reply:
We just check whether that tree path exists in the input_data dict. The type/shape checking is not performed on the value of the input_dict.

Reply:
oh regarding this function, not we don't do anything especial. SpecDict itself handles None values to only check for existence of the tree path.